pkg/datapath/linux/node.go

```diff
@@ -127,6 +129,59 @@ func updateTunnelMapping(oldCIDR, newCIDR *cidr.CIDR, oldIP, newIP net.IP, first
 	}
 }
 
+// updateIPCacheExternalVxlanMapping is called
+// with external VXLAN device encapsulation mode enabled.
+// the CIDR, Endpoint, MAC are of external VXLAN device
+func updateIPCacheExternalVxlanMapping() {
+	var (
+		worldID              uint32 = 2 //identity
+		encryptKey           uint8  = 0 //no encrypt
+		externalVxlanCIDR    *cidr.CIDR
+		exernalVxlanEndpoint net.IP
+		externalVtepMAC      mac.MAC
+		err                  error
+	)
+
+	for i := 0; i < len(option.Config.ExternalVxlanEndpoint); i++ {
+		externalVxlanCIDR, err = cidr.ParseCIDR(option.Config.ExternalVxlanCIDR[i])
+		if err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				"allocCIDR": option.Config.ExternalVxlanCIDR[i],
+			}).Error("bpf: Unable to parse external VXLAN CIDR")
+			continue
+		}
+
+		exernalVxlanEndpoint = net.ParseIP(option.Config.ExternalVxlanEndpoint[i]).To4()
+		if exernalVxlanEndpoint == nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				"allocEndpoint": option.Config.ExternalVxlanEndpoint[i],
+			}).Error("bpf: Unable to external VXLAN endpoint")
+			continue
+		}
+
+		externalVtepMAC, err = mac.ParseMAC(option.Config.ExternalVxlanMAC[i])
+		if err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				"allocMAC": option.Config.ExternalVxlanMAC[i],
+			}).Error("bpf: Unable to parse external VXLAN MAC")
+			continue
+		}
+		log.WithFields(logrus.Fields{
+			logfields.IPAddr: exernalVxlanEndpoint,
+			"allocCIDR":      externalVxlanCIDR,
+		}).Info("Updating ipcache map entry")
+
+		err = ipcacheMap.UpdateIPCacheMapping(externalVxlanCIDR, exernalVxlanEndpoint, worldID, externalVtepMAC, encryptKey)
```

Comment:
Is this bypassing `pkg/ipcache`? I think that the ipcache update should go through there to ensure that we consistently resolve potential conflicts between different sources of ipcache data, and to ensure that users/developers can introspect the ipcache using `cilium ip list ...`.

Reply:
This should also help with getting the identity correct for these ipcache entries. Rather than directly injecting `worldID`, this should ideally use `ipcache.AllocateCIDRsForIPs()` to ensure that the identity is resolved consistently for these VTEPs.

Reply:
I think that this is pretty nuanced, partly because of the way Cilium is configured and the way it handles identities (and in particular, identities for CIDRs) internally.

The external VTEP should be configured to send VNI 2 (world).

When receiving packets from a VTEP (ie, the MAC is nonzero), we should ensure that the VNI is `ID_WORLD`. This prevents traffic from being received into the datapath policy processing and being treated with higher privilege than it should.

The ipcache should use CIDR identity allocation (which is local to the agent) inside the ipcache. This has implications on the way that `toCIDR` policy is handled internally in Cilium, like for instance you can write policies for your pods to allow traffic to `10.0.0.0/16` and this can allow the pod to send to any IPs in that range. If your VTEP is inside that range, then the toCIDR /16 policy should also allow traffic towards that destination. If the identity for the VTEP IP is set directly to `2` then this policy will not work. If the identity is set via CIDR identity allocation, then the CIDR policy will work.

Reply:
> I think that this is pretty nuanced, partly because of the way Cilium is configured and the way it handles identities (and in particular, identities for CIDRs) internally.
> 
> The external VTEP should be configured to send VNI 2 (world).
> 
> When receiving packets from a VTEP (ie, the MAC is nonzero), we should ensure that the VNI is `ID_WORLD`. This prevents traffic from being received into the datapath policy processing and being treated with higher privilege than it should.
> 

Yes, I agree above, I have made changes to address above.

> The ipcache should use CIDR identity allocation (which is local to the agent) inside the ipcache. This has implications on the way that `toCIDR` policy is handled internally in Cilium, like for instance you can write policies for your pods to allow traffic to `10.0.0.0/16` and this can allow the pod to send to any IPs in that range. If your VTEP is inside that range, then the toCIDR /16 policy should also allow traffic towards that destination. If the identity for the VTEP IP is set directly to `2` then this policy will not work. If the identity is set via CIDR identity allocation, then the CIDR policy will work.

Are you suggesting to let ipcache allocate CIDR and identity for VTEP? not use world id `2` for VTEP if I want cilium pod policy  work for VTEP? I am not familiar with pkg/ipcache, and  how cilium allocate each podCIDR for each node, I took a quick look at pkg/ipcache, not understanding how it works,  I have a few questions which may make no sense since I am  foreign to the ipcache/CIDR/identity/policy... :)

1,  ipcache can allocate CIDR outside of cluster podCIDR range for example `cluster-pool-ipv4-cidr: "10.0.0.0/8"`, if so how? I thought ipcache can only allocate CIDR within cluster podCIDR range for example `10.0.0.0/8`, or is it ipcache allocate podCIDR for each node?

2 if ipcache can allocate CIDR for VTEP, I can update the ipcache with VTEP CIDR with `vtep_mac`?  

3, how do I make sure the VTEP CIDR allocated by ipcache does not change for whatever reasons and I don't need to re-configure the VTEP device

I think you raise a good point on the policy use case for VTEP,  it is beyond my understanding  at this point in time, I prefer to make this PR as simple as I can understand it, when I get more in-depth understanding cilium policy and implementation, I can improve this in the future :)